### PR TITLE
feat: add `foreach (x <- xs; y <- ys) exp` expression

### PR DESCRIPTION
**Pull request by jaschdoc**
Related to #3950

Is there a better title for this PR?

### CODE REVIEW

**Comment by jaschdoc** [comment]
RFR

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 1]
```diff
@@ -2156,4 +2155,15 @@ object ParsedAst {
     * Represents an optional purity and optional effect.
     */
   case class PurityAndEffect(pur: Option[Type], eff: Option[EffectSet])
+
+  /**
+    * A foreach expression generator, i.e. `x <- xs`.
+    *
+    * @param sp1 the position of the first character in the expression.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:2162]: not an expression :)

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 2]
```diff
@@ -2156,4 +2155,15 @@ object ParsedAst {
...
+    * @param pat the pattern on the left hand side.
+    * @param exp the iterable expression.
+    * @param sp2 the position of the last character in the expression.
+    */
+  case class ForEachGenerator(sp1: SourcePosition, pat: ParsedAst.Pattern, exp: ParsedAst.Expression, sp2: SourcePosition)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:2167]: We can use this name, until we come up with a better one.

I guess we will have more of these when we add the other for loop.

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 3]
```diff
@@ -502,8 +502,12 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
       SP ~ keyword("discard") ~ WS ~ Expression ~ SP ~> ParsedAst.Expression.Discard
     }
 
+    def ForEachGenerator: Rule1[ParsedAst.ForEachGenerator] = rule {
+      SP ~ Pattern ~ WS ~ keyword("<-") ~ WS ~ Expression ~ SP ~> ParsedAst.ForEachGenerator
+    }
+
     def ForEach: Rule1[ParsedAst.Expression.ForEach] = rule {
-      SP ~ keyword("foreach") ~ optWS ~ "(" ~ optWS ~ Pattern ~ WS ~ keyword("<-") ~ WS ~ Expression ~ optWS ~ ")" ~ optWS ~ Expression ~ SP ~> ParsedAst.Expression.ForEach
+      SP ~ keyword("foreach") ~ optWS ~ "(" ~ optWS ~ oneOrMore(ForEachGenerator).separatedBy(optWS ~ ";" ~ optWS) ~ optWS ~ ")" ~ optWS ~ Expression ~ SP ~> ParsedAst.Expression.ForEach
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:510]: Do we want to the last whitespace to be hard or is it ok to write `foreach(...)123+2`?

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 4]
```diff
@@ -502,8 +502,12 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
       SP ~ keyword("discard") ~ WS ~ Expression ~ SP ~> ParsedAst.Expression.Discard
     }
 
+    def ForEachGenerator: Rule1[ParsedAst.ForEachGenerator] = rule {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:505]: This can be nested inside ForEach since it is not use anywhere else.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:505]: (Ask Matt if you get stuck)

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala [hunk 5]
```diff
@@ -687,23 +687,25 @@ object Weeder {
...
-      val exp2Val = visitExp(exp2, senv)
-      val fqn = "ForEach.foreach"
 
-      mapN(patVal, exp1Val, exp2Val) {
-        case (p, e1, e2) =>
-          val lambda = mkLambdaMatch(sp1, p, e2, sp2)
-          mkApplyFqn(fqn, List(lambda, e1), loc)
+      val fqn = "ForEach.foreach"
+      val loc = mkSL(sp1, sp2).asSynthetic
+      val patVals = traverse(gens)(g => visitPattern(g.pat))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:697]: traverse(gens)(visitPattern(_.pat)

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala (hunk 3 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Parser.scala:510]: I personally think we shouldn't be too restrictive on the white space unless there's a really good reason to.

Some C programmers for example write their stuff without white space hehe :wink:

File: main/src/ca/uwaterloo/flix/language/phase/Weeder.scala (hunk 5 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:697]: Doesn't compile with this... for some reason it only accepts `g => visitPattern(g.pat)`

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Weeder.scala:697]: Ok

**Pull request merged**